ext/pdo_sqlite/pdo_sqlite_class.c

```diff
@@ -0,0 +1,657 @@
+
+	if ((ret = zend_call_function(&fc->fci, &fc->fcc)) == FAILURE) {
+		php_error_docref(NULL, E_WARNING, "An error occurred while invoking the callback");
+	}
+
+	/* clean up the params */
+	if (zargs) {
+		for (i = is_agg; i < fake_argc; i++) {
+			zval_ptr_dtor(&zargs[i]);
+		}
+		if (is_agg) {
+			zval_ptr_dtor(&zargs[1]);
+		}
+		efree(zargs);
+	}
+
+	if (!is_agg || !argv) {
+		/* only set the sqlite return value if we are a scalar function,
+		 * or if we are finalizing an aggregate */
+		if (!Z_ISUNDEF(retval)) {
+			switch (Z_TYPE(retval)) {
+				case IS_LONG:
+					sqlite3_result_int(context, Z_LVAL(retval));
+					break;
+
+				case IS_NULL:
+					sqlite3_result_null(context);
+					break;
+
+				case IS_DOUBLE:
+					sqlite3_result_double(context, Z_DVAL(retval));
+					break;
+
+				default:
+					if (!try_convert_to_string(&retval)) {
+						ret = FAILURE;
+						break;
+					}
+					sqlite3_result_text(context, Z_STRVAL(retval), Z_STRLEN(retval), SQLITE_TRANSIENT);
+					break;
+			}
+		} else {
+			sqlite3_result_error(context, "failed to invoke callback", 0);
+		}
+
+		if (agg_context) {
+			zval_ptr_dtor(&agg_context->val);
+		}
+	} else {
+		/* we're stepping in an aggregate; the return value goes into
+		 * the context */
+		if (agg_context) {
+			if (Z_ISUNDEF(retval)) {
+				zval_ptr_dtor(&agg_context->val);
+				return FAILURE;
+			}
+			zval_ptr_dtor(Z_REFVAL(agg_context->val));
+			ZVAL_COPY_VALUE(Z_REFVAL(agg_context->val), &retval);
+			ZVAL_UNDEF(&retval);
+		}
+	}
+
+	if (!Z_ISUNDEF(retval)) {
+		zval_ptr_dtor(&retval);
+	}
+
+	return ret;
+}
+
+
+void php_pgsql_func_callback(sqlite3_context *context, int argc,
+	sqlite3_value **argv)
+{
+	struct pdo_sqlite_func *func = (struct pdo_sqlite_func*)sqlite3_user_data(context);
+
+	do_callback(&func->afunc, &func->func, argc, argv, context, 0);
+}
+
+static void php_pgsql_func_step_callback(sqlite3_context *context, int argc,
+	sqlite3_value **argv)
+{
+	struct pdo_sqlite_func *func = (struct pdo_sqlite_func*)sqlite3_user_data(context);
+
+	do_callback(&func->astep, &func->step, argc, argv, context, 1);
+}
+
+static void php_pgsql_func_final_callback(sqlite3_context *context)
+{
+	struct pdo_sqlite_func *func = (struct pdo_sqlite_func*)sqlite3_user_data(context);
+
+	do_callback(&func->afini, &func->fini, 0, NULL, context, 1);
+}
+
+
+
+
+
+
+/* {{{ proto bool PdoSqlite::createFunction(string $function_name, callable $callback, int $num_args = -1, int $flags = 0)
+    Creates a function that can be used in a query
+*/
+PHP_METHOD(PdoSqlite, createFunction)
+{
+	//copied from sqlite_driver.c
+
+	struct pdo_sqlite_func *func;
+	zend_fcall_info fci;
+	zend_fcall_info_cache fcc;
+	char *func_name;
+	size_t func_name_len;
+	zend_long argc = -1;
+	zend_long flags = 0;
+	pdo_dbh_t *dbh;
+	pdo_sqlite_db_handle *H;
+	int ret;
+
+	ZEND_PARSE_PARAMETERS_START(2, 4)
+		Z_PARAM_STRING(func_name, func_name_len)
+		Z_PARAM_FUNC(fci, fcc)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG(argc)
```

Comment:
Probably want to add a check and ValueError for when this is negative

Reply:
Done.